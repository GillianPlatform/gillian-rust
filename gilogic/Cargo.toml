[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }