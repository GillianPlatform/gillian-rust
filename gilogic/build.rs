fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
