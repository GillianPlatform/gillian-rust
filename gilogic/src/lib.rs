#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
