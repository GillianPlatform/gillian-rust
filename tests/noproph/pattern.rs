//@check-pass
extern crate gilogic;

use gilogic::{macros::*, Ownable};

#[specification(
  requires { emp }
  ensures { emp }
)]
fn test((a, b): (i32, i32)) -> i32 {
    return 0;
}
