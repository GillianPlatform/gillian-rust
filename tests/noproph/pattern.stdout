#begin_init_data
{}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_no_pcy.gil", "i__lang_no_pcy.gil";

spec test(G_1)
[[ (#G_1 == {{ #a, #b }}) * (G_1 == #G_1) ]]
[[ emp ]]
normal

proc test(G_1) {
        m_1G_1 := G_1;
        g_0 := m_1G_1;
        m_1G_1 := [alloc]({{ "tuple", {{ "i32", "i32" }} }});
        u := [store_value](l-nth(m_1G_1, 0i), l-nth(m_1G_1, 1i), {{ "tuple", {{ "i32", "i32" }} }}, g_0);
  bb0:  g_2 := [load_value](l-nth(m_1G_1, 0i), l+(l-nth(m_1G_1, 1i), {{ {{ "f", 0i, {{ "tuple", {{ "i32", "i32" }} }} }} }}), "i32", true);
        g_1 := l-nth(g_2, 0i);
        m_2a := g_1;
        g_4 := [load_value](l-nth(m_1G_1, 0i), l+(l-nth(m_1G_1, 1i), {{ {{ "f", 1i, {{ "tuple", {{ "i32", "i32" }} }} }} }}), "i32", true);
        g_3 := l-nth(g_4, 0i);
        m_3b := g_3;
        m_0 := 0i;
        g_5 := m_0;
        ret := g_5;
        goto rlab;
  rlab: u := [free](l-nth(m_1G_1, 0i), l-nth(m_1G_1, 1i), {{ "tuple", {{ "i32", "i32" }} }});
        return
};

